feat: DIA-2092:  Support reading multiple tasks from each file in a cloud storage

matt-bernstein opened the pull request:
[x] allow reading multiple tasks from the same file during import storage sync in scan_and_create_links
[x] update get_data for all import storage types to return list[dict] instead of dict
[x] handle blob urls? ~currently, still assuming only 1 task per file using these. Changing this would have a larger blast radius, would have to read the url at import time to determine how many tasks are in it, and then make up a task format for pointing to an object within a url, and read the url n times for n tasks during normal usage unless it's cached somehow...~ use blob urls == "Treat every bucket object as a source file", so it doesn't make sense to do this
[x] do the same for LSE S3 and GCS WIF import storage
[x] add test coverage - currently most test coverage for io_storages is in a single tavern file. Each import storage class definitely needs its own coverage for multiple tasks per key, so this'll be pretty lengthy

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

matt-bernstein committed:
feat: DIA-2092: Support reading multiple tasks from each file in a cloud storage

matt-bernstein committed:
initial implementation for azure

matt-bernstein committed:
add other storage types

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
bugfixes

matt-bernstein committed:
add tests

matt-bernstein committed:
allow code reuse

matt-bernstein committed:
remove unecessary args

pakelley commented on label_studio/io_storages/azure_blob/models.py:
```diff
@@ -122,20 +122,28 @@ def iterkeys(self):
+        elif isinstance(value, list):
+            for idx, item in enumerate(value):
+                if not isinstance(item, dict):
+                    raise ValueError(
```
Unfortunately, I don't think this message will actually reach the user. I think the better behavior may be to log the exception (so we see it in sentry) and import anything we can

pakelley replied:
Worth checking w/ Sheree if she has an opinion here though tbh

matt-bernstein replied:
Check out the existing implementation for s3 - some storages raise an error here and some don't, I chose to add the same error to all of them to keep it consistent. I could be missing some storage-specific logic outside this function that means they're handled differently though.

pakelley commented on label_studio/io_storages/base_models.py:
```diff
@@ -398,26 +398,27 @@ def _scan_and_create_links(self, link_class):
-                )
-                tasks_for_webhook = []
+            for task_data in tasks_data:
+                task = self.add_task(task_data, self.project, maximum_annotations, max_inner_id, self, key, link_class)
```
Mind adding a TODO comment to potentially make an `add_tasks` function so we can batch these?

pakelley reviewed (approved):
I don't see anything blocking, lgtm

matt-bernstein committed:
add comment

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
attempted test isolation debugging

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein committed:
use proper inheritance hierarchy

matt-bernstein committed:
fix tests

matt-bernstein committed:
rename param

matt-bernstein committed:
blue

matt-bernstein committed:
ruff

matt-bernstein committed:
revert ruff

matt-bernstein committed:
fork except on windows

matt-bernstein committed:
blue

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-dia-2092

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14720984501)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14720984501

matt-bernstein committed:
sort imports

matt-bernstein committed:
imports

matt-bernstein committed:
blue

matt-bernstein committed:
fork

matt-bernstein committed:
reorder marks

matt-bernstein committed:
ruff

matt-bernstein committed:
skip fork

matt-bernstein committed:
ruff???

Commit:
ruff

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14722513446)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14722513446

hakan458 commented on label_studio/io_storages/base_models.py:
```diff
@@ -429,26 +429,29 @@ def _scan_and_create_links(self, link_class):
+                # `emit_webhooks_for_instance`, and `tasks_for_webhook` is cleared for new tasks.
+                # If tasks remain in `tasks_for_webhook` at process end (less than `WEBHOOK_BATCH_SIZE`), they're sent in a final webhook
+                # call to ensure all tasks are processed and no task is left unreported in the webhook.
+                if len(tasks_for_webhook) >= settings.WEBHOOK_BATCH_SIZE:
```
This could potentially be huge and I dont know what our limitations are for this webhook but I figure we should batch these to WEBHOOK_BATCH_SIZE as to not send thousands of tasks at once

hakan458 replied:
Oh wait I guess we append one task at a time to the list right, we dont add the entire list of tasks at once, so all good

matt-bernstein replied:
yep, this block is just moved inside a new loop, logic stays the same
